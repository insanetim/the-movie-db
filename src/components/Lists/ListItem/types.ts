import { MouseEvent } from 'react'
import { IList } from 'src/interfaces/list.interface'

export type ListItemHookProps = {
  listId: IList['id']
  name: IList['name']
}

export type ListItemHookReturn = {
  handleClick: () => void
  handleListDelete: (event: MouseEvent<HTMLSpanElement>) => () => Promise<void>
}

export type ListItemProps = {
  description: IList['description']
  listId: IList['id']
  name: IList['name']
}
