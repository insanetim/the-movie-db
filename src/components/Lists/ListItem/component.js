import PropTypes from 'prop-types'
import { Col, Card, Typography } from 'antd'
import { DeleteOutlined } from '@ant-design/icons'

import useContainer from './hook'

const ListItem = ({ list }) => {
  const { handleClick, handleDelete } = useContainer({ listId: list.id })

  return (
    <Col
      span={24}
      sm={{ span: 12 }}
      md={{ span: 8 }}
      lg={{ span: 8 }}
      xl={{ span: 6 }}
    >
      <Card
        className='top-margin'
        hoverable
        actions={[
          <DeleteOutlined
            key='delete'
            onClick={handleDelete}
          />
        ]}
        onClick={handleClick}
      >
        <Typography.Title level={4}>{list.name}</Typography.Title>
        {list.description}
      </Card>
    </Col>
  )
}

ListItem.propTypes = {
  list: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    description: PropTypes.string
  })
}

ListItem.defaultProps = {
  list: {}
}

export default ListItem
