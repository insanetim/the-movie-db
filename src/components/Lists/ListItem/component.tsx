import { DeleteOutlined } from '@ant-design/icons'
import { Card, Typography } from 'antd'

import useContainer from './hook'
import { ListItemProps } from './types'

const ListItem: React.FC<ListItemProps> = ({ description, listId, name }) => {
  const { handleConfirmDeleteList, handleNavigateToList } = useContainer({
    listId,
    name,
  })

  return (
    <Card
      actions={[
        <DeleteOutlined
          data-testid='deleteListBtn'
          key='delete'
          onClick={handleConfirmDeleteList}
        />,
      ]}
      data-testid='listItemCard'
      hoverable
      onClick={handleNavigateToList}
    >
      <Typography.Title level={4}>{name}</Typography.Title>
      {description}
    </Card>
  )
}

export default ListItem
