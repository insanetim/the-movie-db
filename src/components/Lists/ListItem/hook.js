import { Modal } from 'antd'
import { useDispatch } from 'react-redux'
import { useNavigate } from 'react-router-dom'

import { deleteList } from 'src/state/lists/actions'

const useContainer = ({ listId }) => {
  const dispatch = useDispatch()
  const navigate = useNavigate()

  const handleClick = () => navigate(`/list/${listId}`)

  const handleDelete = event => {
    event.stopPropagation()
    const onOk = () => {
      dispatch(deleteList({ listId }))
    }
    Modal.confirm({
      title: 'Do you want to delete list?',
      onOk
    })

    return onOk
  }

  return { handleClick, handleDelete }
}

export default useContainer
