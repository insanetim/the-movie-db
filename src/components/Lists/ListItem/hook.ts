import { Modal } from 'antd'
import { MouseEvent } from 'react'
import { useSelector } from 'react-redux'
import { useNavigate, useSearchParams } from 'react-router-dom'
import { useAppDispatch } from 'src/hooks/useRedux'
import useUpdatePage from 'src/hooks/useUpdatePage'
import { deleteList, fetchLists } from 'src/store/createdLists/actions'
import { createdListsSelector } from 'src/store/createdLists/selectors'
import getSlug from 'src/utils/helpers/getSlug'

import { ListItemHookProps, ListItemHookReturn } from './types'

const useContainer = ({ id, name }: ListItemHookProps): ListItemHookReturn => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const lists = useSelector(createdListsSelector)
  const [searchParams, setSearchParams] = useSearchParams()
  const page = searchParams.get('page') || '1'
  const { updatePage } = useUpdatePage({
    action: fetchLists(page),
    items: lists?.results,
    page,
    setSearchParams,
  })

  const handleClick = () => {
    navigate(`/list/${getSlug(id, name)}`)
  }

  const handleListDelete = (event: MouseEvent<HTMLSpanElement>) => {
    event.stopPropagation()

    const onOk = async () => {
      await dispatch(deleteList(id))
      updatePage()
    }

    Modal.confirm({
      onOk,
      title: 'Do you want to delete list?',
    })

    return onOk
  }

  return { handleClick, handleListDelete }
}

export default useContainer
