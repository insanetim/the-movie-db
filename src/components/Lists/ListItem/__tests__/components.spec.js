import { render } from '@testing-library/react'

import Wrapper from 'src/__mocks__/wrapperMock'
import ListItem from '../component'

const mockedHookData = {
  handleClick: jest.fn(),
  handleDelete: jest.fn()
}
jest.mock('../hook', () => jest.fn(() => mockedHookData))

describe('ListItem component', () => {
  it('matches snapshot', () => {
    const mockedList = {
      id: 1,
      name: 'test/name',
      description: 'test/description'
    }
    const { asFragment } = render(<ListItem list={mockedList} />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
})
