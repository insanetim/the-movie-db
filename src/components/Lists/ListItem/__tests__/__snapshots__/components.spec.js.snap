// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Col
  lg={
    {
      "span": 8,
    }
  }
  md={
    {
      "span": 8,
    }
  }
  sm={
    {
      "span": 12,
    }
  }
  span={24}
  xl={
    {
      "span": 6,
    }
  }
>
  <Card
    actions={
      [
        <ForwardRef(DeleteOutlined)
          onClick={[MockFunction]}
        />,
      ]
    }
    className="top-margin"
    hoverable={true}
    onClick={[MockFunction]}
  >
    <ForwardRef
      level={4}
    >
      Test list
    </ForwardRef>
    Test description
  </Card>
</Col>
`;
